File: python/0234_Palindrome_Linked_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        
        def middleNode(head):
            fast = slow = head
            while fast and fast.next:
                fast = fast.next.next
                slow = slow.next
            return slow
        
        def reverseList(head):
            prev = None
            while head:
                nextNode = head.next
                head.next = prev
                prev = head
                head = nextNode
            return prev
        
        middle = middleNode(head)
        second = reverseList(middle)
        first = head
        while second:
            if first.val != second.val:
                return False
            first = first.next
            second = second.next
        return True

# O(N) Time
# O(1) Space

File: cpp/0234_Palindrome_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* fast = head;
        ListNode* slow = head;
        while (fast && fast->next) {
            fast = fast->next->next;
            slow = slow->next;
        }
        return slow;
    }
    ListNode* reverseList(ListNode* head) {
        ListNode *dummy = new ListNode(-1);
        dummy->next = head;
        while (head && head->next) {
            ListNode* nextNode = head->next;
            head->next = head->next->next;
            nextNode->next = dummy->next;
            dummy->next = nextNode;
        }
        return dummy->next;
    }
    bool isPalindrome(ListNode* head) {
        ListNode* middle = middleNode(head);
        ListNode* second = reverseList(middle);
        ListNode* first = head;
        while (second) {
            if (first->val != second->val) return false;
            first = first->next;
            second = second->next;
        }
        return true;
    }
};

// O(N) Time
// O(1) Space